feat: create cron nft-ttr to measure nft time to retrievability

gobengo opened the pull request:
Motivation:
* https://github.com/nftstorage/nft.storage/issues/1943

Plan

1. note all feedback
    * dont use jest for cron tests. use ava
    * dont use yargs for argv parsing, use `"sade": "^1.7.4"`
    * use less fields on nft-ttr metric [via](https://filecoinproject.slack.com/archives/C02BZPRS9HP/p1654044830889049?thread_ts=1653585353.059339&cid=C02BZPRS9HP)
    * make nft-ttr retrieval metric a histogram instead of a gauge [via](https://filecoinproject.slack.com/archives/C02BZPRS9HP/p1654055779880559?thread_ts=1653585353.059339&cid=C02BZPRS9HP)
    * alan [thoughts](https://github.com/nftstorage/nft.storage/issues/1943#issuecomment-1143650993)
        * It would be nice to be uploading some useful data rather than just junk
        * A typical NFT use case 10k drops
            * Folks typically upload a directory with 10k JSON files and also 10k of images (the files must not be the same or deduplication will occur)
        * We don't have to upload 10MB every time. In the Web3.Storage client we expose the DAG splitting size as a param to put. In NFT.Storage you could use carbites to split the DAG by a smaller size and then call storeCar multiple times.
        * When available (https://github.com/ipfs/go-ipfs/issues/8640), you might want to use the ?format=car parameter on the gateways to verify ALL the blocks of the DAG are accessible.
4. updates based on review/blocking feedback
1. merge this - whatever is required to get approval
2. prioritized other tasks to do while awaiting review approval
    * [x] update this PR to use sade for arg parsing
    * [x] update this PR such that cron package uses ava for test running (no jest)
    * [x] test retrieving from more than one gateway
6. verify once merged
    * crons are running by gh actions once the workflow is described in main branch
    * crons result in metrics successfully pushed to pushgateway
7. followup pr with
    * fixes if the aforementioned verifications fail

Possible Futures
* @alanshaw [requirement 'We should also record the time it took to upload the data.'](https://github.com/nftstorage/nft.storage/pull/1945#discussion_r892340411)
  * https://github.com/nftstorage/nft.storage/issues/1994
* upgrade tsc https://github.com/nftstorage/nft.storage/pull/1945#issuecomment-1155533297
* eslint rest of files and simplify eslint config exceptions https://github.com/nftstorage/nft.storage/pull/1945#discussion_r892271164
* [ ] change nft-ttr so it doesn't generate 10Mb every time, it just generates ~1Mb but triggers at least one car split by manually configuring the client
  * [ ] blocked by https://github.com/nftstorage/nft.storage/issues/1962 
* [ ] propose [ADRs](https://adr.github.io/)
* [ ] propose ADR in nft.storage for using sade for arg parsing
* [ ] propose ADR in nft.storage for using ava as test runner
* add test coverage requirements to cron usage of c8 (e.g. thresholds from [ucanto](https://github.com/web3-storage/ucanto/blob/main/package.json#L26))
* make this job have the nft.storage-specific parts be injectable so same job can be used to observe web3.storage
* disable eslint overrides (which were added to avoid unnecessarily modifying files otherwise not touched by this PR) in packages/cron/package.json `.eslintConfig.overrides`
* add another cronjob to continuously test @alanshaw's [thought](https://github.com/nftstorage/nft.storage/issues/1943#issuecomment-1143650993)
  > * A typical NFT use case 10k drops
  >   * Folks typically upload a directory with 10k JSON files and also 10k of images (the files must not be the same or deduplication will occur)

gobengo committed:
add .github/workflows/cron-nft-ttr

gobengo committed:
nft-ttr uses yargs to accept --url flag defaulting to nft.storage

gobengo committed:
nft-ttr downloads file to use as example image when measuring

gobengo committed:
nft-ttr downloads a source file to use for rest of test

gobengo committed:
measureNftTimeToRetrievability.js logs duration

yusefnapora committed:
fix: preserve all backup_urls for chunked uploads (#1940)

* fix: preserve all backup_urls for chunked uploads

* rm stray file

* respond to review feedback

gobengo committed:
measureNftTimeToRetrievability.js pushes metrics to metricsGateway

gobengo committed:
measureNftTimeToRetrievability.spec.js test doesnt hit network, asserts log of retrieval duration

gobengo committed:
Add github workflow to typecheck/test cron

gobengo committed:
lint cron

gobengo committed:
cron typechecks

gobengo committed:
github workflow for cron check does npm-install in cron dir

gobengo committed:
try new workflow cron check install-command

gobengo committed:
dont use gozala typecheck action

gobengo committed:
gh cron log

gobengo committed:
cron debug

gobengo committed:
client has prepack script

gobengo committed:
cron debug

gobengo committed:
cron debug

gobengo committed:
cron rm mime-types

gobengo committed:
cron debug

gobengo committed:
cron debug

gobengo commented on packages/cron/src/jobs/measureNftTimeToRetrievability.js:
```diff
@@ -0,0 +1,215 @@
+    }
+    config.log('info', storeLog)
+    const retrievalUrl = new URL(
+      `https://${metadata.ipnft}.ipfs.nftstorage.link/image/blob`
```
eventually should test a handful of gateways, not just this one. perhaps as separate job invocations (e.g. using github actions matrix)

gobengo replied:
https://github.com/nftstorage/nft.storage/pull/1945/commits/176ebeab8fd1a665010290404db039f1b8644661

gobengo committed:
measureNftTimeToRetrievability can be tested with >10Mb files

gobengo committed:
move online test to nft-ttr.spec.js

gobengo committed:
update cron-nft-ttr.yml

gobengo committed:
nft-ttr support retrieving from multiple gateways

gobengo committed:
fix cron typecheck

gobengo committed:
.github cron-nft-ttr invokes using multiple gateways

gobengo committed:
github cron-nft-ttr passes secrets.PUSHGATEWAY_BASIC_AUTH

gobengo commented on .github/workflows/cron-nft-ttr.yml:
```diff
@@ -0,0 +1,36 @@
+          DEBUG: '*'
+          ENV: ${{ matrix.env }}
+          NFT_STORAGE_API_KEY: ${{ secrets.NFT_STORAGE_API_KEY }}
+          PUSHGATEWAY_BASIC_AUTH: ${{ secrets.PUSHGATEWAY_BASIC_AUTH }}
```
I haven't been able to test the pathway when this is truthy because I don't have a sample secret

gobengo replied:
nvm I found a value in secrets vault and tested locally. then made small fixes until the push metrics request responded with status 200

gobengo committed:
measureNftTimeToRetrievability.js metrics pushgateway works locally with sample value from secrets vault

gobengo committed:
rm default metricsPushGateway and provide it in gh action

gobengo commented:
looks like some of my sample metrics pushes show up [here](https://protocollabs.grafana.net/explore?left=%7B%22datasource%22:%22Prometheus-k8s%22,%22queries%22:%5B%7B%22exemplar%22:true,%22expr%22:%22nftstorage_retrieval_duration_seconds%22,%22interval%22:%22%22,%22queryType%22:%22randomWalk%22,%22refId%22:%22A%22,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%22000000044%22%7D%7D%5D,%22range%22:%7B%22from%22:%22now-24h%22,%22to%22:%22now%22%7D%7D&orgId=1)

yusefnapora committed:
chore: refactor service configuration loading (#1914)

* chore: refactor service configuration loading

Refactors the way we load configuration variables. Key changes:

- constants.js is replaced with config.js
- config.js exports `getServiceConfiguration`, which lazily loads
the config into a JS `ServiceConfiguration` object
- default values for unit tests are defined in code, in config.js
- runtime values are dynamically looked up on `globalThis`, to avoid
undefined reference errors if a var is missing.
- if a config variable doesn't exist at runtime, we log a warning and
use the default value

All existing tests pass, but the config code itself isn't tested yet.

* embrace truthiness

* add test for serviceConfigFromVariables

* fancier bool parsing

* add tests for loading config vars & throw on missing vars in prod/staging

* add hack to workaround global VERSION issue

* include all missing vars in error / warning message

* flatten out config object & simplify default value fallback

* fixup for prev commit: get maintenance mode from config

* chore: add namespace to VERSION, etc

* fix maintenence mode tests

* rm dev cruft

* fix rebase cruft

gobengo committed:
yarn.lock

gobengo committed:
ghwf cron-nft-ttr invokes with camelCase flag not kebab-case

gobengo committed:
ci(cron): action doesnt explicitly check for NFT_STORAGE_API_KEY env

gobengo commented on .github/workflows/cron-nft-ttr.yml:
```diff
@@ -0,0 +1,38 @@
+          yarn workspace --silent cron run --silent start:nft-ttr measure \
+            --min-image-size-bytes=10000000 \
+            --gateways https://nftstorage.link https://dweb.link \
+            --metricsPushGateway $PUSHGATEWAY_URL \
```
no reason, but I changed it to be consistent: https://github.com/nftstorage/nft.storage/pull/1945/commits/b7d958b35fd4f7b1b4e727cf57fdc02e1dd33794

gobengo committed:
rename BYTES_10_MB -> defaultTestMinImageSizeBytes

address https://github.com/nftstorage/nft.storage/pull/1945#discussion_r886878216

gobengo committed:
add descriptive variable name isProcessEntrypoint in bin/nft-ttr.js

gobengo commented on packages/cron/src/bin/nft-ttr.spec.js:
```diff
@@ -0,0 +1,43 @@
+import { recordedLog } from '../lib/log'
+import * as assert from 'node:assert'
+
+const BYTES_10_MB = 10 * 1e6
```
ok https://github.com/nftstorage/nft.storage/pull/1945/commits/c27d2299b236ba84d0bf142644f5471c04e4fd74

gobengo committed:
rm unused UrlIamages

gobengo committed:
rename fn TestImages -> createTestImages

gobengo committed:
createTestImages uses RandomImageBlob

gobengo commented on packages/cron/src/jobs/measureNftTimeToRetrievability.js:
```diff
@@ -0,0 +1,278 @@
+      yield resp.blob()
+    }
+  })()
+}
```
> How come these functions begin with uppercase? We usually reserve for classes or react components.

no good reason. I'll use that convention going forward

gobengo replied:
also added `@typescript-eslint/naming-convention` warning so lint catches this

gobengo committed:
packages/cron uses @typescript-eslint/recommended and naming-convention

jsdevel committed:
feat: Adding HasDeleteRestriction user_tag (#1952)

* Adding the type and failing HTTP DELETE operations if this tag is set.
* See nftstorage/admin.storage#66

gobengo committed:
measureNftTimeToRetrievability.js merges config & { secrets } into options

gobengo commented on packages/cron/src/jobs/measureNftTimeToRetrievability.js:
```diff
@@ -0,0 +1,278 @@
+ * @param {string} secrets.nftStorageToken - API Token for nft.storage
+ * @param {HttpAuthorization} secrets.metricsPushGatewayAuthorization - authorization header value for metricsPushGateway
+ */
+export async function measureNftTimeToRetrievability(config, secrets) {
```
I like to try to isolate secrets early so they dont actually get logged to ci stdout for all to see.
e.g. I didn't want to afford for users of the script to pass secrets via cli flags, which would then show up in process command. I wanted them to pass using a separate channel from 'config' (e.g. env vars).
Not that attached to having them separate at this layer.
I did something to merge them into a shared options object.

gobengo committed:
add Milliseconds.subtract static

gobengo commented on packages/cron/src/jobs/measureNftTimeToRetrievability.js:
```diff
@@ -0,0 +1,278 @@
+      token: secrets.nftStorageToken,
+    })
+    const storeStartedDate = new Date()
+    const storeStartedAt = now()
```
imho passing around an untyped int is more complexity than this because then whoever's dealing with the code later will have to remember the units instead of the compiler doing it for us.

gobengo committed:
measureNftTimeToRetrievability checks success via response.ok and not status allowList

addresses https://github.com/nftstorage/nft.storage/pull/1945/files/efa794b23197f36ecad68c93f377e7a2e59dc9d3#diff-70187cabafb845fa5768abccc9ef51efd84529fb89e9f37188058f90fde3d69d

gobengo committed:
cron eslintConfig rm unneeded .excludedFiles

gobengo committed:
dont eslint-disable cron jobs/dagcargo , disable via eslintConfig overrides instead

gobengo committed:
start integrating prom-client

gobengo committed:
refactor metrics

gobengo committed:
progress on metrics

github-actions[bot] committed:
chore(main): release api 2.24.2 (#1930)

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

jsdevel committed:
feat: DB schema and API for user_tag_proposal. (#1410) (#1402)

* Users create records in this table and admins manage it.

Co-authored-by: trigramdev9 <jsdevel@trigram.co>

hugomrdias committed:
fix: config loading required vars (#1955)

gobengo committed:
measureNftTimeToRetrievability test works on airplane with no wifi

gobengo committed:
nft-ttr uses httpImageFetcher

gobengo committed:
measureNftTimeToRetrievability.spec asserts that pushRetrievalMetrics is called

gobengo committed:
cleanup cron nftttr

gobengo committed:
cron uses sade instead of yargs

gobengo committed:
cron tests work offline by stubbing network calls

gobengo committed:
cron package uses ava instead of jest for testing

gobengo commented on packages/cron/src/jobs/measureNftTimeToRetrievability.js:
```diff
@@ -0,0 +1,278 @@
+    [
+      '# TYPE nftstorage_retrieval_duration_seconds gauge',
+      '# HELP nftstorage_retrieval_duration_seconds How long it took to retrieve an nft image',
+      `nftstorage_retrieval_duration_seconds{image="${retrieval.image}",url="${retrieval.url}",size="${retrieval.contentLength}"} ${durationSeconds}`,
```
I switched to prom-client

gobengo commented on packages/cron/src/jobs/measureNftTimeToRetrievability.js:
```diff
@@ -0,0 +1,278 @@
+  ]
+  const pushResponse = await fetch.apply(null, pushRequest)
+  const pushSuccessfulStatusCodes = [200, 201]
+  if (!pushSuccessfulStatusCodes.includes(pushResponse.status)) {
```
done

gobengo committed:
measureNftTimeToRetrievability and nft-ttr return AsyncIterable and tests assert that instead of recorded info logs

gobengo committed:
cron eslint uses plugin:prettier/recommended

lanzafame commented on packages/cron/src/jobs/measureNftTimeToRetrievability.js:
```diff
@@ -0,0 +1,380 @@
+    [
+      '# TYPE nftstorage_retrieval_duration_seconds gauge',
+      '# HELP nftstorage_retrieval_duration_seconds How long it took to retrieve an nft image',
+      `nftstorage_retrieval_duration_seconds{byteLength="${retrieval.contentLength}"} ${durationSeconds}`,
```
@gobengo You need to move to a histogram. A gauge is the wrong measurement type for this metric. I outlined my recommendation in the slack thread on this PR but for convenience, tldr, remove the url and image field, `size` should be a set of buckets, i.e. 10MB, 25MB, 50MB, 100MB and the metric should be a histogram (https://www.npmjs.com/package/prom-client#Histogram)

gobengo replied:
@lanzafame Thanks for the review. Helped me realize this code is now unused, as I switched to using npm `prom-client` per @alanshaw suggestion, and in doing so I switched to a [Histogram](https://github.com/nftstorage/nft.storage/pull/1945/files#diff-5823c1dd943cae9a4760e898922c5e7a0344a231d51560db2541882f2a9beb41R15).

> size should be a set of buckets, i.e. 10MB, 25MB, 50MB, 100MB

By size do you mean `byteLength`? fwiw I removed all mention of the fields. I think we're mostly interested in measuring the retrieval duration in seconds. The byteLength/size was a nice to have.

Thanks for educating me about the perf et al tradeoffs in storing extra fields and not using buckets.

lanzafame replied:
If the bytelength is still variable (as in the size of the file uploaded does change) then it should be recorded against the duration value so that the histogram can be presented correctly on the grafana end. Otherwise we will have uploads that take longer due to the file size but appear to be because the service is being slow.

gobengo committed:
cron eslint add plugin:@typescript-eslint/recommended-requiring-type-checking

gobengo committed:
ci: cron.yml prepares client before linting cron

gobengo committed:
remove unused code pushing to prom pushgateway via fetch (and using wrong metric type)

gobengo committed:
remove unused byteLength field from createRetrievalDurationSecondsMetric

gobengo committed:
cron test uses c8 to report code coverage

gobengo commented on packages/cron/src/jobs/measureNftTimeToRetrievability.spec.js:
```diff
@@ -0,0 +1,58 @@
+      }
+    )
+    assert.equal(storeSpy.mock.calls.length, 1)
+    const start = info.find((logs) => logs[0]?.type === 'start')[0]
```
Good point. I switched to AsyncIterable

gobengo committed:
cron remove unused safe-env-vars dep

gobengo committed:
cron remove unused dep debug

gobengo committed:
cron eslint remove unused not-from-recommended rules

gobengo committed:
cron eslint remove unused eslintIgnore values

gobengo committed:
cron remove unused eslint-disable file comments

gobengo commented:
@alanshaw I believe I've addressed much of your feedback from last review. wdyt?

jsdevel committed:
feat: Adding admin ability to search by github_id (#1960)

* See nftstorage/admin.storage#68

github-actions[bot] committed:
chore(main): release api 2.25.0 (#1956)

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

johnathonroach committed:
feat: optimize metrics query (#1906)

This commit optimizes the metrics query to be less brittle
in the case that the dimensions array data changes or
takes on new values. Here, we query the presence
of the keys instead of the exact match of the
stringified array.

github-actions[bot] committed:
chore(main): release api 2.26.0 (#1966)

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

flea89 committed:
feat: cron release workflow (#1932)

github-actions[bot] committed:
chore(main): release cron 3.1.0 (#1968)

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

alanshaw commented on packages/cron/package.json:
```diff
@@ -11,29 +11,170 @@
+  },
+  "eslintIgnore": [
+    "node_modules"
+  ]
```
Please can you follow up with a PR to fix the linter issues and remove all this configuration? 🙏🙏🙏

gobengo replied:
sure thing. didn't want to lint those old files prematurely until you weighed in on the lint rules at hand

gobengo replied:
@alanshaw this would do it. Not sure if you want to review in here or out. https://github.com/gobengo/nft.storage/pull/1

alanshaw commented on packages/cron/src/lib/metrics.js:
```diff
@@ -0,0 +1,21 @@
+export function createRetrievalDurationSecondsMetric(registry) {
+  /** @type RetrievalDurationSecondsMetric} */
+  const metric = new Histogram({
+    name: 'retrieval_duration_seconds',
```
We should also record the time it took to upload the data.

gobengo replied:
I agree. Can we do it as fast-follow after merge or should I do before?

alanshaw replied:
Yes can be follow up

gobengo commented on packages/cron/src/bin/nft-ttr.js:
```diff
@@ -0,0 +1,280 @@
+        ...options,
+      })
+    })
+  sadeParseWithoutExit(argParser, argv)
```
I tried that but it still exited :/

gobengo commented on packages/cron/src/bin/nft-ttr.js:
```diff
@@ -0,0 +1,280 @@
+ * @param {sade.Sade} prog
+ * @param {string[]} argv
+ */
+function sadeParseWithoutExit(prog, argv) {
```
will do

> It's also a bit unexpected to have a "main" function in a binary that's an async iterable...

I can rename the function

gobengo replied:
I made a 'cli' function that returns the iterable, and the `main` function returns a promise of its doneness.

johnathonroach committed:
feat: add NFTUp CTA to stats page (#1971)

github-actions[bot] committed:
chore(main): release website 1.52.0 (#1975)

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

gobengo committed:
clean createMeasureSecretsFromEnv based on review

lanzafame commented on packages/cron/src/lib/metrics.js:
```diff
@@ -0,0 +1,21 @@
+ * @param {Registry} registry
+ * @returns {RetrievalDurationSecondsMetric}
+ */
+export function createRetrievalDurationSecondsMetric(registry) {
```
This should be milliseconds instead of seconds, unless I missed a sneaky conversion somewhere, milliseconds is the unit that is used everywhere else. Also, the conversion is handled automatically in grafana.

gobengo replied:
from https://prometheus.io/docs/practices/naming/

> ...should use base units (e.g. seconds, bytes, meters - not milliseconds, megabytes, kilometers). See below for a list of base units.

There was a conversion, but using the prom-client `Histogram` directly as a metric wouldn't have included very good type-checking so I added some types to help. https://github.com/nftstorage/nft.storage/pull/1945/commits/440e07e9885bf9544b59b99ac62aebc466ba7ec3#diff-5823c1dd943cae9a4760e898922c5e7a0344a231d51560db2541882f2a9beb41R12

jsdevel committed:
fix: Adding user_tag_proposal schema to reset.sql for local dev (#1977)

* Introduced in #1402

Commit:
docs: add store from URL example (#1979)

Adds an example showing how to store ERC-721 compatible metadata with image data retrieved from a URL on the internet.

Adapted from https://github.com/nftstorage/nft.storage/blob/main/examples/client/browser/store.html

gobengo committed:
nft-ttr error throws with cause instead of console.warn

Co-authored-by: Alan Shaw <alan.shaw@protocol.ai>

gobengo committed:
Revert "nft-ttr error throws with cause instead of console.warn"

This reverts commit 726a7a4d67bd3f78d5e76a984f8f32fd30376152.

gobengo committed:
nft-ttr rename main -> cli

gobengo committed:
nft-ttr tests pass without process.env needing NFT_STORAGE_API_KEY

gobengo committed:
cron: better typechecking around RetrievalDurationMetric value

gobengo committed:
test createMeasureOptionsFromSade

gobengo committed:
test createMeasureSecretsFromEnv

gobengo committed:
rm sadeParseWithoutExit

gobengo committed:
get rid of LogFunction

gobengo commented on packages/cron/src/lib/log.js:
```diff
@@ -0,0 +1,56 @@
+    const stringifiedLoggables = loggables.map((o) => JSON.stringify(o))
+    return log(level, ...stringifiedLoggables)
+  }
+}
```
nope i removed it

gobengo committed:
Merge branch 'main' into issues/1943-measure-ttr

gobengo committed:
test cron utils so coverage is green in c8

gobengo committed:
rm --silent from cron-nft-ttr gh wf

gobengo commented on .github/workflows/cron-nft-ttr.yml:
```diff
@@ -0,0 +1,38 @@
+          PUSHGATEWAY_BASIC_AUTH: ${{ secrets.PUSHGATEWAY_BASIC_AUTH }}
+        run: |
+          yarn workspace --silent cron run --silent start:nft-ttr measure --logConfigAndExit
+          yarn workspace --silent cron run --silent start:nft-ttr measure \
```
got rid of --silent and ndjson support

gobengo committed:
remove unneeded eslint-disable-next-line

gobengo committed:
remove commented console.log

gobengo committed:
Add explanation comment to hasOwnProperty

gobengo commented:
@alanshaw Should I merge this diff that edits unrelated modules to pass new lint rules into this PR? Or do as followup to this one? I'd slightly prefer the latter.
https://github.com/gobengo/nft.storage/pull/1/files

alanshaw commented on package.json:
```diff
@@ -1,8 +1,17 @@
+      "@types/mocha",
+      "@types/node",
+      "@types/webpack-env"
+    ]
```
Why this `nohoist` change? Unless it causing an issue, can we move to a separate PR?

gobengo replied:
It was causing an issue I believe. I will verify it's really required now.

gobengo replied:
it wasn't required. :/ ty

alanshaw reviewed (approved):
Lets get this merged and do the follow up items.

Note: needs a rebase (cron jobs are now integrated with release please so have to be release in the same way as other packages)

gobengo committed:
rm unnecessary nohoists
